sdk/servicebus/azservicebus/processor.go

```diff
@@ -0,0 +1,395 @@
+		return nil
+	}
+}
+
+func ProcessorWithQueue(queue string) ProcessorOption {
+	return func(processor *Processor) error {
+		processor.config.Entity.Queue = queue
+		return nil
+	}
+}
+
+func ProcessorWithSubscription(topic string, subscription string) ProcessorOption {
+	return func(processor *Processor) error {
+		processor.config.Entity.Topic = topic
+		processor.config.Entity.Subscription = subscription
+		return nil
+	}
+}
+
+// ProcessorWithAutoComplete enables or disables auto-completion/abandon of messages
+// When this option is enabled the result of the `processMessage` handler determines whether
+// the message is abandoned (if an `error` is returned) or completed (if `nil` is returned).
+// This option is enabled, by default.
+func ProcessorWithAutoComplete(enableAutoCompleteMessages bool) ProcessorOption {
+	return func(processor *Processor) error {
+		processor.config.ShouldAutoComplete = enableAutoCompleteMessages
+		return nil
+	}
+}
+
+// ProcessorWithMaxConcurrentCalls controls the maximum number of message processing
+// goroutines that are active at any time.
+// Default is 1.
+func ProcessorWithMaxConcurrentCalls(maxConcurrentCalls int) ProcessorOption {
+	return func(processor *Processor) error {
+		processor.config.MaxConcurrentCalls = maxConcurrentCalls
+		return nil
+	}
+}
+
+type legacyNamespace interface {
+	NewReceiver(ctx context.Context, entityPath string, opts ...internal.ReceiverOption) (*internal.Receiver, error)
+}
+
+func newProcessor(ns legacyNamespace, options ...ProcessorOption) (*Processor, error) {
+	processor := &Processor{
+		config: processorConfig{
+			ReceiveMode:        ReceiveModePeekLock,
+			ShouldAutoComplete: true,
+			MaxConcurrentCalls: 1,
+			RetryOptions: struct {
+				Times int
+				Delay time.Duration
+			}{
+				// TODO: allow these to be configured.
+				Times: 10,
+				Delay: time.Second * 5,
+			},
+		},
+
+		mu:                &sync.Mutex{},
+		activeReceiversWg: &sync.WaitGroup{},
+
+		ns:        ns,
+		subscribe: subscribe,
+	}
+
+	for _, opt := range options {
+		if err := opt(processor); err != nil {
+			return nil, err
+		}
+	}
+
+	entityPath, err := formatEntity(processor.config.Entity)
+
+	if err != nil {
+		return nil, err
+	}
+
+	processor.config.FullEntityPath = entityPath
+
+	processor.processorCtx, processor.cancelProcessor = context.WithCancel(context.Background())
+	processor.receiversCtx, processor.cancelReceivers = context.WithCancel(context.Background())
+
+	return processor, nil
+}
+
+// Start will start receiving messages from the queue or subscription.
+//
+//   if err := processor.Start(messageHandler, errorHandler); err != nil {
+//     log.Fatalf("Processor failed to start: %s", err.Error())
+//   }
+//
+//   <- processor.Done()
+//
+// Any errors that occur (such as network disconnects, failures in handleMessage) will be
+// sent to your handleError function. The processor will retry and restart as needed -
+// no user invention is required.
```

Comment:
should this read "no user intervention is required"?

Reply:
Yes, that makes more sense! :)